### PR TITLE
Change random server picking logic to require "general" category, prefer no approval and own language code

### PR DESCRIPTION
**Pull request by kimar**
# Rationale

The following logic is desired:

```
general purpose (must), instant sign-up (should), matching language (should)
```

### CODE REVIEW

File: Mastodon/Scene/Onboarding/PickServer/MastodonPickServerViewModel.swift [hunk 1]
```diff
@@ -235,9 +235,37 @@ extension MastodonPickServerViewModel {
...
+        
+        let randomServer: Mastodon.Entity.Server?
+        
+        let noApprovalRequired = generalServers.filter { !$0.approvalRequired }
+        let approvalRequired = generalServers.filter { $0.approvalRequired }
+        
+        let languageMatchesWithoutApproval = noApprovalRequired.filter { $0.language.lowercased() == language }
+        let languageMatchesWithApproval = approvalRequired.filter { $0.language.lowercased() == language }
+        let languageDoesNotMatchWithoutApproval = noApprovalRequired.filter { $0.language.lowercased() != language }
+        let languageDoesNotMatchWithApproval = noApprovalRequired.filter { $0.language.lowercased() != language }
```
Comment (zeitschlag) [inline Mastodon/Scene/Onboarding/PickServer/MastodonPickServerViewModel.swift]: Those are the same:

```swift
noApprovalRequired.filter { $0.language.lowercased() != language }
noApprovalRequired.filter { $0.language.lowercased() != language }
```

From my understanding (and variable-names) I'd have suspected something like this:

```swift
noApprovalRequired.filter { $0.language.lowercased() != language }
approvalRequired.filter { $0.language.lowercased() != language }
```

File: Mastodon/Scene/Onboarding/PickServer/MastodonPickServerViewModel.swift [hunk 2]
```diff
@@ -235,9 +235,37 @@ extension MastodonPickServerViewModel {
...
+
+        switch (
+            languageMatchesWithoutApproval.isEmpty,
+            languageMatchesWithApproval.isEmpty,
+            languageDoesNotMatchWithoutApproval.isEmpty,
+            languageDoesNotMatchWithApproval.isEmpty
+        ) {
```
Comment (zeitschlag) [inline Mastodon/Scene/Onboarding/PickServer/MastodonPickServerViewModel.swift:257]: I already said it, but: I like this tuple/pattern-matching-approach 👍

**Review by zeitschlag — CHANGES_REQUESTED** [review]
See first comment :)

File: Mastodon/Scene/Onboarding/PickServer/MastodonPickServerViewModel.swift [hunk 3]
```diff
@@ -235,9 +235,37 @@ extension MastodonPickServerViewModel {
...
+            randomServer = languageDoesNotMatchWithApproval.randomElement()
+        case (true, true, false, _):
+            randomServer = languageDoesNotMatchWithoutApproval.randomElement()
+        case (true, false, _, _):
+            randomServer = languageMatchesWithApproval.randomElement()
+        case (false, _, _, _):
+            randomServer = languageMatchesWithoutApproval.randomElement()
+        }
 
         return randomServer ?? servers.randomElement() ?? servers.first
```
Comment (zeitschlag) [inline Mastodon/Scene/Onboarding/PickServer/MastodonPickServerViewModel.swift:270]: If everything goes wrong, this fallback could still lead people to a small instance, but from my perspective, this is not very likely.

File: Mastodon/Scene/Onboarding/PickServer/MastodonPickServerViewModel.swift (hunk 1 above)
Comment (DK1DA) [reply Mastodon/Scene/Onboarding/PickServer/MastodonPickServerViewModel.swift]: Yep you're right, thanks you! fixed.

File: Mastodon/Scene/Onboarding/PickServer/MastodonPickServerViewModel.swift (hunk 3 above)
Comment (DK1DA) [reply Mastodon/Scene/Onboarding/PickServer/MastodonPickServerViewModel.swift:270]: Yes I think that's expected in the worst case

**Review by zeitschlag — APPROVED** [review]
Thank you!

**Pull request merged**